### PR TITLE
Statically linking CUDA for Anaconda builds

### PR DESCRIPTION
**Pull request by pjh5**

### CODE REVIEW

**Comment by pjh5** [comment]
Will manually only run conda tests

File: cmake/Utils.cmake [hunk 1]
```diff
@@ -216,3 +216,33 @@ function(pycmd outvar cmd)
...
+
+###
+# Helper function to print out everything that cmake knows about a target
+#
+# Copied from https://stackoverflow.com/questions/32183975/how-to-print-all-the-properties-of-a-target-in-cmake
+# This isn't called anywhere, but it's very useful when debugging cmake
+# NOTE: This doesn't work for INTERFACE_LIBRARY or INTERFACE_LINK_LIBRARY targets
+
+# Get all propreties that cmake supports
+execute_process(COMMAND cmake --help-property-list OUTPUT_VARIABLE CMAKE_PROPERTY_LIST)
```
Comment (orionr) [inline cmake/Utils.cmake]: Likely all platforms will be fine with this, but, to be safe, I think we should wrap this in a function and cache the result to be used in the `print_target_properties` function below. For example

```
function(populate_cmake_property_list)
    if (!${CMAKE_PROPERTY_LIST})
        execute_process(COMMAND cmake --help-property-list OUTPUT_VARIABLE CMAKE_PROPERTY_LIST)
        ...
    endif()
endfunction()

function(print_target_properties tgt)
    populate_cmake_property_list()
    ... the rest of the function ...
endfunction(print_target_properties)
```

Thoughts?

File: cmake/public/cuda.cmake [hunk 2]
```diff
@@ -89,6 +94,10 @@ find_library(CUDA_NVRTC_LIB nvrtc
     PATH_SUFFIXES lib lib64 lib/x64)
 
 # Create new style imported libraries.
+# Several of these libraries have a hardcoded path if CAFFE2_STATIC_LINK_CUDA
+# is set. This path is where sane CUDA installations have their static
+# libraries installed. Users may have strange CUDA installations, but they
+# should never have this flag set and so shouldn't run into problems.
```
Comment (orionr) [inline cmake/public/cuda.cmake]: Nit: maybe say "this flag should only be used for binary builds, so end users should never have this flag set"

**Review by orionr — APPROVED** [review]
One real comment about moving the `execute_process(COMMAND cmake ...)` into a function that is called lazily. Looks great, though!

**Pull request merged**